src/wal.h

```diff
@@ -25,131 +25,190 @@
-int sqlite3WalSnapshotCheck(Wal *pWal, sqlite3_snapshot *pSnapshot);
-void sqlite3WalSnapshotUnlock(Wal *pWal);
+  int (*xSnapshotGet)(Wal *pWal, sqlite3_snapshot **ppSnapshot);
+  void (*xSnapshotOpen)(Wal *pWal, sqlite3_snapshot *pSnapshot);
+  int (*xSnapshotRecover)(Wal *pWal);
+  int (*xSnapshotCheck)(Wal *pWal, sqlite3_snapshot *pSnapshot);
+  void (*xSnapshotUnlock)(Wal *pWal);
 #endif
 
 #ifdef SQLITE_ENABLE_ZIPVFS
-/* If the WAL file is not empty, return the number of bytes of content
-** stored in each frame (i.e. the db page-size when the WAL was created).
-*/
-int sqlite3WalFramesize(Wal *pWal);
+  /* If the WAL file is not empty, return the number of bytes of content
+  ** stored in each frame (i.e. the db page-size when the WAL was created).
+  */
+  int (*xFramesize)(Wal *pWal);
+#endif
+
+  /* Return the sqlite3_file object for the WAL file */
+  sqlite3_file *(*xFile)(Wal *pWal);
+
+#ifdef SQLITE_ENABLE_SETLK_TIMEOUT
+  int (*xWriteLock)(Wal *pWal, int bLock);
 #endif
 
-/* Return the sqlite3_file object for the WAL file */
-sqlite3_file *sqlite3WalFile(Wal *pWal);
+  void (*xDb)(Wal *pWal, sqlite3 *db);
 
+  const char *zName;
+  struct libsql_wal_methods *pNext;
+} libsql_wal_methods;
+
+libsql_wal_methods* libsql_wal_methods_find(const char *zName);
+
+/* Object declarations */
+typedef struct WalIndexHdr WalIndexHdr;
+typedef struct WalIterator WalIterator;
+typedef struct WalCkptInfo WalCkptInfo;
+
+
+/*
+** The following object holds a copy of the wal-index header content.
+**
+** The actual header in the wal-index consists of two copies of this
+** object followed by one instance of the WalCkptInfo object.
+** For all versions of SQLite through 3.10.0 and probably beyond,
+** the locking bytes (WalCkptInfo.aLock) start at offset 120 and
+** the total header size is 136 bytes.
+**
+** The szPage value can be any power of 2 between 512 and 32768, inclusive.
+** Or it can be 1 to represent a 65536-byte page.  The latter case was
+** added in 3.7.1 when support for 64K pages was added.  
+*/
+struct WalIndexHdr {
+  u32 iVersion;                   /* Wal-index version */
+  u32 unused;                     /* Unused (padding) field */
+  u32 iChange;                    /* Counter incremented each transaction */
+  u8 isInit;                      /* 1 when initialized */
+  u8 bigEndCksum;                 /* True if checksums in WAL are big-endian */
+  u16 szPage;                     /* Database page size in bytes. 1==64K */
+  u32 mxFrame;                    /* Index of last valid frame in the WAL */
+  u32 nPage;                      /* Size of database in pages */
+  u32 aFrameCksum[2];             /* Checksum of last frame in log */
+  u32 aSalt[2];                   /* Two salt values copied from WAL header */
+  u32 aCksum[2];                  /* Checksum over all prior fields */
+};
+
+/*
+** An open write-ahead log file is represented by an instance of the
+** following object.
+*/
+struct Wal {
```

Comment:
Would it be possible to make this opaque for upper layers (Pager)? It seems that a lot of details are being exposed. Can this become a problem for some very exotic WAL implementations?

Reply:
Pager doesn't assume anything about struct Wal, it used to be opaque before this patch. To be honest, I don't remember why this struct is itself exposed, as we should only care about exposing libsql_wal_methods to the user. Perhaps it's an artifact from previous iterations that should just be removed

Reply:
Ok, I refreshed my knowledge a little bit - wal.h is still a header private to libSQL files and no module assumes anything about it. In particular, pager.c does not include it. It is however needed for people who implement their own WAL routines, because their own implementations need to know the WAL structure, that's why this definition is exposed in a header.